### PR TITLE
Rewrite of Progress Bar Graph

### LINKED ISSUE #2992: spinners aren't contained in progres bar on bottom of the screen
in places where spinners should be seen on progres bar there's a blank space
good visible in start of this beatmap: https://osu.ppy.sh/beatmapsets/323961#osu/719902
logs:
[database.log](https://github.com/ppy/osu/files/2181325/database.log)
[network.log](https://github.com/ppy/osu/files/2181326/network.log)
[performance.log](https://github.com/ppy/osu/files/2181327/performance.log)
[performance-gamethread.audio.log](https://github.com/ppy/osu/files/2181328/performance-gamethread.audio.log)
[performance-gamethread.draw.log](https://github.com/ppy/osu/files/2181329/performance-gamethread.draw.log)
[performance-gamethread.input.log](https://github.com/ppy/osu/files/2181330/performance-gamethread.input.log)
[performance-gamethread.update.log](https://github.com/ppy/osu/files/2181331/performance-gamethread.update.log)
[runtime.log](https://github.com/ppy/osu/files/2181332/runtime.log)

### PR DESCRIPTION
**Pull request by ticotaco72**
It' still a work in progress
- Closes #2992 

Changing the way how progress bar is generated during gameplay.

### CODE REVIEW

**Comment by ticotaco72** [comment]
aren't we deleting some HitObjects here?
https://github.com/ppy/osu/blob/1796ffde14021c3cdb5203ddcc0e4437099ca490/osu.Game/Screens/Play/SquareGraph.cs#L130

**Comment by ticotaco72** [comment]
i've asked some questions on issue page.
i'll be very happy if someone could help me with this things.

**Comment by ticotaco72** [comment]
Why appveyor can't build it would when it can?

**Comment by ticotaco72** [comment]
Is there some important reason why granularity is set 200?
https://github.com/ppy/osu/blob/1796ffde14021c3cdb5203ddcc0e4437099ca490/osu.Game/Screens/Play/SongProgressGraph.cs#L22

**Comment by FreezyLemon** [comment]
@ticotaco72 appveyor does not only check if the code can compile, it also checks certain style rules that are defined for osu.
If you look at the appveyor log, it fails at a step "inspectcode" which is exactly this style check. You can go to the "Messages" tab to see what exactly went wrong, or you can inspect your code locally ([check here for more info](https://www.jetbrains.com/help/resharper/InspectCode.html))

**Comment by TPGPL** [comment]
That's not how you make comments

**Comment by ticotaco72** [comment]
Sorry, i maked them to save this places for me. I will delete them when i end work on this pr

**Comment by ticotaco72** [comment]
it's some kind of strange but i have 812 build errors about input in this branch.
?????

**Comment by HoutarouOreki** [comment]
You have to update nuggets

**Comment by ticotaco72** [comment]
big thanks :)

**Comment by ticotaco72** [comment]
are foreach structures permitted? everywhere i write foreach i get build error

**Comment by HoutarouOreki** [comment]
You can find the answer yourself if you try to search for the `foreach` keyword throughout the repo. Also learn to read error messages, they're there for a reason.

**Comment by ticotaco72** [comment]
but look that tells me to do sometihing like this 
`foreach (OsuDifficultyHitObject h )in difficultyBeatmap)`

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 1]
```diff
@@ -23,6 +23,68 @@ public OsuDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
...
+        public List<double> OsuDifficultyAt (IBeatmap beatmap, double timeRate)
+        {
+            //throws list of difficulties(strains) at sections (or at hitobjects)
+            //remember that first section is double-timed
+
+            //if (!beatmap.HitObjects.Any())
+            //    return new List<double>();
+
+            //private OsuDifficultyBeatmap difficultyBeatmap = new OsuDifficultyBeatmap(beatmap.HitObjects.Cast<OsuHitObject>().ToList(), timeRate);
+            //Skill[] skills =
```
Comment (jorolf) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: You forgot ``new Skill[]``

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 2]
```diff
@@ -23,6 +23,68 @@ public OsuDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
...
+            // The first object doesn't generate a strain, so we begin with an incremented section end
+            //double currentSectionEnd = 2 * sectionLength;
+
+            //foreach ( OsuDifficultyHitObject h in difficultyBeatmap )
+            //{
+            //    while (h.BaseObject.StartTime > currentSectionEnd)
+            //    {
+            //        foreach (Skill s in skills)
+            //        {
+            //            s.SaveCurrentPeak()
```
Comment (jorolf) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: You forgot a semicolon here

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 3]
```diff
@@ -23,6 +23,68 @@ public OsuDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
...
+            //        }
+
+            //        currentSectionEnd += sectionLength;
+            //    }
+
+            //    foreach (Skill s in skills)
+            //        s.Process(h);
+            //}
+
+            //private List<double> aimRating = new List<double>();
```
Comment (jorolf) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: Why ``private``? You can't that for variable declarations inside methods

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 4]
```diff
@@ -23,6 +23,68 @@ public OsuDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
...
+            //private List<double> speedRating = new List<double>();
+            //private List<double> osuDifficultySectionRating = new List<double>();
+
+            //foreach(double x in Skills[0].StrainPeaks)
+            //{
+            //    aimRating.Add(x * difficulty_multiplier);
+            //};
```
Comment (jorolf) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: Remove this semicolon

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 5]
```diff
@@ -23,6 +23,68 @@ public OsuDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
         {
         }
 
+        public OsuDifficultySectionRating  = new List<double>();
```
Comment (jorolf) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: You forgot to specify the type of this field. And why is it public?

**Review by jorolf — CHANGES_REQUESTED** [review]

**Comment by ticotaco72** [comment]
thanks

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs (hunk 5 above)
Comment (ticotaco72) [reply osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: i thinked about it as output of function at some time

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 6]
```diff
@@ -23,6 +24,66 @@ public OsuDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
...
+        public List<double> OsuDifficultyAt (IBeatmap beatmap, double timeRate)
+        {
+            //throws list of difficulties(strains) at sections (or at hitobjects)
+            //remember that first section is double-timed
+
+            if (!beatmap.HitObjects.Any())
+                return new List<double>();
+
+            private OsuDifficultyBeatmap difficultyBeatmap = new OsuDifficultyBeatmap(beatmap.HitObjects.Cast<OsuHitObject>().ToList(), timeRate);
+            new Skill[] skills =
```
Comment (jorolf) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: ``Skill[] skills = new Skill[]``

File: osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs [hunk 7]
```diff
@@ -23,6 +24,66 @@ public OsuDifficultyCalculator(Ruleset ruleset, WorkingBeatmap beatmap)
...
+                    }
+
+                    currentSectionEnd += sectionLength;
+                }
+
+                foreach (Skill s in skills)
+                    s.Process(h);
+            }
+
+            var List<double> aimRating = new List<double>(), speedRating = new List<double>(), osuDifficultySectionRating = new List<double>();
```
Comment (jorolf) [inline osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs]: please learn c# 🤦‍♂️

**Comment by ticotaco72** [comment]
thanks a lot @jorolf, and sorry for making problems by my c# knowledge

**Comment by peppy** [comment]
Please make your code compile before PRing.

**Pull request closed without merging**